core/src/main/java/com/linecorp/armeria/internal/client/DefaultClientRequestContext.java

```diff
@@ -849,4 +850,33 @@ private String toStringSlow(@Nullable Channel ch, @Nullable RequestLogAccess par
             return buf.toString();
         }
     }
+
+    @Override
+    public CompletableFuture<Void> initiateConnectionShutdown() {
+        final CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+
+        setAdditionalRequestHeader(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE);
+        log().whenRequestComplete().handle((ignore, ex) -> {
+            if (ex != null) {
+                completableFuture.completeExceptionally(ex);
+                return null;
+            }
+            final Channel ch = channel();
+            if (ch == null) {
+                completableFuture.complete(null);
+                return null;
+            }
```

Comment:
- `Channel` is acquired first and the request could complete exceptionally later. How about checking first?
- I don't see if we can complete `completableFuture` successfully if `ch` is null. As `initiateConnectionShutdown` fails to shut down, we need to complete it exceptionally. 	How about creating an `IllegalStateException("the request has failed even before a connection is established.")`?

Reply:
I understand that we complete it exceptionally with `IllegalStateException("a request has failed before a connection is established.")`, when `ch` is null and `ex` is not null. We should check firstly that `ch` is null.
Is it right?

Reply:
`ex` can be used only when `ch` is null. Otherwise, we can successfully complete the future. 
For example:
```java
if (ch == null) {
   if (ex == null) {
     ex = new IllegalStateException(..);
   }
   completableFuture.completeExceptionally(ex);
} else {
   ch.closeFuture().addListener(f -> ...);
}
```

Reply:
Ah, I see.